text/0000-grid-primitives.md

```diff
@@ -0,0 +1,680 @@
+
+- If a Schema already exists with this name or the name is an empty string, the
+  transaction is invalid.
+- If the property list is empty, the transaction is invalid.
+- The signer of the transaction must be an agent in Pike state and must belong
+  to an organization in Pike state, otherwise the transaction is invalid.
+- The agent must have the permission `can_create_schema` for the organization,
+  otherwise the transaction is invalid.  
+
+The schema is created with the provided fields, in addition to the Pike
+organization ID as the `owner_id`. The schema is then stored in state,
+
+The inputs for SchemaCreateAction must include:
+
+- Address of the Agent submitting the transaction
+- Address of the Schema
+  
+The outputs for SchemaCreateAction must include:
+
+- Address of the Schema  
+
+#### SchemaUpdateAction
+
+SchemaUpdateAction updates a Schema to state. This update only adds new
+Properties to the Schema.
+
+```
+message SchemaUpdateAction {
+    string schema_name = 1;
+    repeated PropertyDefinition properties = 2;
+}
+```
+
+The action is validated according to the following rules:
+
+- If a Schema does not exist, the transaction is invalid.
+- If the property list is empty, the transaction is invalid.
+- If one of the new properties has the same name as a property already defined
+  in the scheme, the  transaction is invalid.
+- The signer of the transaction must be an agent in the Pike state and must
+  belong to an organization in Pike state, otherwise the transaction is invalid.
+- The signer of the transaction must belong to the same organization matching
+  the `owner` of the schema, otherwise the transaction is invalid.
+- The agent must have the permission `can_update_schema` for the organization,
+  otherwise the transaction is invalid.
+
+The inputs for SchemaUpdateAction must include:
+
+- Address of the Agent submitting the transaction
+- Address of the Schema  
+
+The outputs for SchemaCreateAction must include:
+
+- Address of the Schema  
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+In this RFC, validating a collection of property values in a data structure
+against a Schema is left to the smart contract implementer. This may result in
+inconsistent validation rules, particularly in the cases where the natural
+default value provided by protobuf deserialization may be ambiguous (for
+example, the enum case).
+
+The dynamic nature of the property values also require a query for the schema in
+order to validate the state.  This incurs an additional cost when creating and
+updating a collection of property values. 
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+This design provides a reusable pattern for arbitrary properties in data
+structures defined in Grid smart contracts. Providing a common language across
+all data structures built with the schemas will allow for the creation of a
+common set of tools, such as REST APIs and Sawtooth State Delta Export.
```

Comment:
"state delta export" should not have initial caps (to be consistent with the Sawtooth documentation and writing best practices).

Reply:
Fixed